File: src/add_two_ints_server.cpp
#include <ros/ros.h>
#include <bingda_prictices/addtwoints.h>//包含服务消息头文件


bool add(bingda_prictices::addtwoints::Request &req,
         bingda_prictices::addtwoints::Response &res)//服务回调函数
{
  res.sum = req.a + req.b;
  ROS_INFO("request: x=%ld, y=%ld", (long int)req.a, (long int)req.b);
  ROS_INFO("sending back response: [%ld]", (long int)res.sum);
  return true;
}


int main(int argc, char **argv)
{
  ros::init(argc, argv, "add_two_ints_server");
  ros::NodeHandle n;

  ros::ServiceServer service = n.advertiseService("add_two_ints", add);//创建一个服务服务器，服务名称为"add_two_ints"，回调函数为add
  ROS_INFO("Ready to add two ints.");//输出日志信息
  ros::spin();//循环等待回调函数

  return 0;
}

File: scripts/add_two_ints_server.py
#!/usr/bin/env python3
import rospy
from std_msgs.msg import String
from bingda_prictices.msg import student  # 导入我们自己定义的消息类型
from bingda_prictices.srv import addtwoints, addtwointsResponse  # 导入我们自己定义的服务类型和服务响应类型

def handle_add_two_ints(req):
    sum = req.a + req.b
    rospy.loginfo("Adding %d and %d, result is %d", req.a, req.b, sum)
    return addtwointsResponse(sum)  # 返回结果

def add_two_ints_server():
    rospy.init_node('add_two_ints_server')  # 初始化ROS节点
    s = rospy.Service('add_two_ints', addtwoints, handle_add_two_ints)  # 创建服务，服务名称为add_two_ints，服务类型为addtwoints，回调函数为handle_add_two_ints
    rospy.loginfo("Ready to add two ints.")  # 打印日志信息
    rospy.spin()  # 保持节点运行

if __name__ == "__main__":
    add_two_ints_server()  # 启动服务服务器
